### PR TITLE
4662 om supply backups

### LINKED ISSUE #4662: omSupply backups
## Is your feature request related to a problem? Please describe 👀

Implement backup and restore functionality as per: https://github.com/msupply-foundation/open-msupply/pull/3924

## Describe the solution you'd like 🎁

Extend existing omSupply cli to copy app_data and backup database (copy sqlite files for sqlite, and do pg_dump for postgres). Existing cli already uses configuration files, and makes it easier to locate app_data + database location, port, credentials etc..

### Describe alternatives you've considered 💭

https://github.com/msupply-foundation/open-msupply/pull/3924

### PR DESCRIPTION
**Pull request by andreievg**
Fixes #4662 

Original PR is here, please make sure all comments are address in that PR as well: https://github.com/msupply-foundation/open-msupply/pull/4664

# 👩🏻‍💻 What does this PR do?

Adds restore and backup cli scripts, please see [readme](https://github.com/msupply-foundation/open-msupply/blob/81092014a6efa0421953027cb011073969f446b2/server/cli/src/backup/README.md) for more details

I originally had backup folder and pg_bin_dir as cli arguments, after talking to Kahn we thought it would be best in configurations

## 💌 Any notes for the reviewer?

There are bunch of carry overs and improvements that we should consider doing, i'll try to relocate to another issue:

There is `database_path` settings, which i think is not used (it's not used for sqlite i think, but makes more sense for sqlite?), if it's used for sqlite need to adjust cli -> It is used, so would be good for backup to respect that setting, not deemed high priority because central server is the only real use case for production backup and it should be running in postgres

Currently copying all folders in app_data, if there was an easier way to use the files service for this, i think it would be preferred, but would require a bit of refactor, I didn't want to touch it

Database drop and creation can be made to be re-usable, but i didn't want to touch core code

Would be great to specify configuration files when launching server or running backup script, this would make running central and remote server a bit easier.

Add some metadata to backup, is it sqlite or postgres, which version, what client/site etc.. 

I've noticed awhile ago that database errors formatting is not that great, I think we can improve on this by using this error on RepositoryError.

Auto zip backup

# 🧪 Testing

For both sqlite and postgres versions:

- [ ] Backup (1)
- [ ] Make some changes (add invoice/requisition)
- [ ] Backup again (2)
- [ ] Delete some things (invoice/requisition)
- [ ] Backup again (3)
- [ ] Restore (1) make sure invoice/requisition does not exist, and creating new invoice has sequence number + 1
- [ ] Restore (2) make sure invoice/requisition exists, and creating new invoice has sequence number + 1
- [ ] Restore (3) make sure invoice/requisition does not exist, and creating new invoice has sequence number + 2

Check that files are also backed up (for best done I think with assets)

# 📃 Documentation

Does this need docs beyond docs in the repo ?

### CODE REVIEW

File: server/cli/src/backup/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+`NOTE` In production backups should only be done for omSupply central servers, remote sites are backed up via synchronisation
+
+### Backup
+
+To backup simply run: 
+
+**In development mode**
+
+`cargo run --bin remote service_cli -- backup`
```
Comment (CarlosNZ) [inline server/cli/src/backup/README.md:13]:
```suggestion
cargo run --bin remote_server_cli -- backup
```

I think this is what you mean, right? (And for the other ones too)

**Comment by CarlosNZ** [comment]
> Would be great to specify configuration files when launching server or running backup script, this would make running central and remote server a bit easier.

What I did for the [database monitor](https://github.com/msupply-foundation/database-monitor) is offer a [separate `config` parameter](https://github.com/msupply-foundation/database-monitor?tab=readme-ov-file#load-configuration) where you can specify the path to the config file to use. You only need to run it once, as it stores the location for future commands. This means users can store the config file anywhere they like, but without the hassle of having to specify it for every command.

**Comment by CarlosNZ** [comment]
I'm having trouble getting this to run. I'm getting this error:
```
Error: Cannot find sqlite backup files with name omsupply-db.sqlite
```

My `omsupply-db.sqlite` file is in its normal location, and it definitely exists!

I presume we should be calling this script from within the `/server` folder? (It failed when I tried anything else)

I added the `backup` settings to my `local.yaml` file as suggested, and I tried adding a `database_path` property as well (although I wasn't clear on what that's supposed to do based on the above comments)

Also -- how do you specify whether to back up the SQLite file or the Postgres database?

**Comment by mark-prins** [comment]
> how do you specify whether to back up the SQLite file or the Postgres database?

if you build with `--features postgres` it will do the pg version. probably should say that in the readme right?

**Comment by andreievg** [comment]
> What I did for the [database monitor](https://github.com/msupply-foundation/database-monitor) is offer a [separate config parameter](https://github.com/msupply-foundation/database-monitor?tab=readme-ov-file#load-configuration) where you can specify the path to the config file to use. You only need to run it once, as it stores the location for future commands. This means users can store the config file anywhere they like, but without the hassle of having to specify it for every command.

That's great, there is slightly different use case for the config in omSup (actually it is overlapping a little bit), we would want to launch omSupply central and omSupply remote from the repo. So command being remembered is not super helpful for that use case (otherwise config lives in relative location to executable, so no real need to remember it), i think the overlap is in that you may want to run multiple instances of database monitor at the same time with different configs (i think at this point remembering the config might not be that useful ?).

Thanks for the heads up though, I think it would be applicable (actually, if multiple configs have been used, would be great if it showed the list to pick from !)

**Comment by andreievg** [comment]
> My omsupply-db.sqlite file is in its normal location, and it definitely exists!

Sorry for that, i didn't expect the .sqlite to be specified in database_name, fixed, for backwards compatibility: https://github.com/msupply-foundation/open-msupply/pull/4674/commits/f67168c68543cb9389b5676773bf31c2a16abf3a

> Also -- how do you specify whether to back up the SQLite file or the Postgres database?
> if you build with --features postgres it will do the pg version. probably should say that in the readme right?

Thanks @mark-prins, also updated here: 63439207f4ef2463654a9fbbb3a57416c51e205f

File: server/cli/src/backup/README.md (hunk 1 above)
Comment (andreievg) [reply server/cli/src/backup/README.md]: Yeah, sorry: 63439207f4ef2463654a9fbbb3a57416c51e205f

**Comment by andreievg** [comment]
> Got to the point where it creates a backup.. and I can't find it!

Sorry Mark, I haven't quite tested the `~` tilda in the example (I tested backup and restore but didn't check where it was saving), i added `shellexpand` and it should work now: 7faf4206cee8af66223ab31b860975e9f0f08c35

**Comment by andreievg** [comment]
> minor point: likely more useful to have a windows example as we don't have any macos servers in the wild

noted, will add to carry over or sneak a commit in, just need to check if %homepath% will work (will check after next build)

File: server/cli/src/backup/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+Backup will contain a folder with all of the app_data (plugins, static files, etc..) and a folder with either sqlite files or postgres database dump. 
+
+### Restore
+
+To restore run: 
+
+**In development mode**
+
+`cargo run --bin remote_server_cli -- restore -b D2024_08_22T05_05_16 -s` or for postgres `cargo run --bin remote_server_cli --features postgres -- restore -b D2024_08_22T05_05_16 -s`
```
Comment (CarlosNZ) [inline server/cli/src/backup/README.md:29]: Also, do you think you should explain what the `-s` flag does? I realised it was for "skip confirmation" with some trial and error, but might be good to be thorough here.

Or... can the flag be more explicit? (i.e `-skip-confirmation`)?

**Review by CarlosNZ — APPROVED** [review]
Looks good, working well now, thanks.

A few tiny usability-related comments. And just one more: can you print some kind of "Success" message to the console when the restore is finished? Currently there's nothing, so it's hard to be confident it's worked.

In the future, might be good to have a "zip" option. If we want to share snapshots around (which I certainly will), we'll probably manually zip them first anyway, so would be a bit convenient to have this baked in. But that can wait :)

Cheers

File: server/cli/src/backup/backup.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    copy_files(settings, &file_dir)?;
+
+    // Backup database
+    if cfg!(feature = "postgres") {
+        dump_postgres_database(settings, &database_dir, pg_bin_dir)?;
+    } else {
+        copy_sqlite_files(settings, &database_dir)?;
+    }
+
+    println!("{backup_name}");
```
Comment (mark-prins) [inline server/cli/src/backup/backup.rs:28]: when using interactively a 'backup completed' would be nice
```suggestion
    println!("Backup completed in folder {backup_name}");
```

File: server/README.md [hunk 4]
```diff
@@ -331,8 +331,13 @@ cargo run --bin remote_server_cli -- initialise-from-export -n 'export_name' -r
 cargo run --bin remote_server_cli -- initialise-from-central -u 'user1:password1'
 # attempt to refresh dates (advance them forward, see --help)
 cargo run --bin remote_server_cli -- refresh-dates
+# See README.md link below
+cargo run --bin remote_server_cli -- backup
+cargo run --bin remote_server_cli -- restore D2024_08_22T05_05_16 -s
```
Comment (mark-prins) [inline server/README.md:336]:
```suggestion
cargo run --bin remote_server_cli -- restore --backup-name D2024_08_22T05_05_16 -s
```

or `-b` probs worth mentioning either is possible

File: server/cli/src/backup/restore.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
+use super::*;
+use copy_dir::copy_dir;
+use diesel::{Connection, RunQueryDsl};
+use repository::DBBackendConnection;
+use service::settings::Settings;
+use simple_log::is_debug;
+use std::{fs, io, path::PathBuf, process::Command, str::FromStr};
+
+pub(crate) fn restore(
```
Comment (mark-prins) [inline server/cli/src/backup/restore.rs:9]: thought for later is how we can make the errors look nicer:

```
cargo run --bin remote_server_cli --features postgres -- restore --backup-name D2024_08_26T06_01_27 -s
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.28s
     Running `target/debug/remote_server_cli restore --backup-name D2024_08_26T06_01_27 -s`
Error: "UNKNOWN" ("\"\\\"database \\\\\\\"omsupply\\\\\\\" is being accessed by other users\\\"\"")
```

**Review by mark-prins — APPROVED** [review]
that's very nice! thanks Andrei, great work

File: server/cli/src/backup/restore.rs (hunk 5 above)
Comment (andreievg) [reply server/cli/src/backup/restore.rs:9]: I agree ! I also noticed and added to PR description (to relocate to new issue at some point)

File: server/cli/src/backup/README.md (hunk 2 above)
Comment (andreievg) [reply server/cli/src/backup/README.md]: I was hoping that more detailed and up to date info can be found with `--help`, added: 97a1bc76ada567de0a4e93481496eac8a8e83aea

File: server/cli/src/backup/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+`NOTE` In production backups should only be done for omSupply central servers, remote sites are backed up via synchronisation
+
+### Backup
+
+To backup simply run: 
+
+**In development mode**
+
+`cargo run --bin remote_server_cli -- backup` or for postgres `cargo run --bin remote_server_cli --features postgres -- backup`
```
[the remark this answers is not in the document]
Comment (andreievg) [reply server/cli/src/backup/README.md]: Thanks, i didn't realise that the copy icon exists, will use triple quote from now on: 97a1bc76ada567de0a4e93481496eac8a8e83aea

File: server/README.md (hunk 4 above)
Comment (andreievg) [reply server/README.md]: Thanks updated with `-b` and also the `--help` should show full description (short and long form)

File: server/cli/src/backup/backup.rs (hunk 3 above)
Comment (andreievg) [reply server/cli/src/backup/backup.rs:28]: It looks like support will not be using the stdio, happy with this ! Thanks

**Comment by andreievg** [comment]
> In the future, might be good to have a "zip" option. If we want to share snapshots around (which I certainly will), we'll probably manually zip them first anyway, so would be a bit convenient to have this baked in. But that can wait :)

Thanks added to PR comments, will relocate to another issue

**Comment by jmbrunskill** [comment]
FWIW, SQLite recommends doing a `vacuum into` or backup api to backup when using the WAL. 
https://www.sqlite.org/lang_vacuum.html
https://www.sqlite.org/backup.html

**Comment by andreievg** [comment]
Thanks @jmbrunskill, I've added to the PR description (to be transferred to carry over issue)

```
Emphasise that sqlite backup is for dev use only (backup in production should only be done on central server which should always run in postgres flavour)
```

**Pull request merged**